src/app/components/hooks/users.ts

```diff
@@ -0,0 +1,83 @@
+import { auth, db } from "@/lib/firebase";
+import { User as FirebaseUser, getAuth } from "firebase/auth";
+import { User } from "@/types/user";
+import {
+  collection,
+  doc,
+  getDoc,
+  getDocs,
+  updateDoc,
+} from "firebase/firestore";
+
+export const getUser = async (): Promise<User | null> => {
+  return new Promise<User | null>((resolve, reject) => {
+    const unsubscribe = auth.onAuthStateChanged(
+      async (firebaseUser: FirebaseUser | null) => {
+        if (firebaseUser) {
+          try {
+            const userDocRef = doc(db, "users", firebaseUser.uid);
+            const userDoc = await getDoc(userDocRef);
+            if (userDoc.exists()) {
+              const userData = userDoc.data();
+              const mappedUser: User = {
+                uid: firebaseUser.uid,
+                displayName:
+                  userData.displayName || firebaseUser.displayName || undefined,
+                email: userData.email || firebaseUser.email || "",
+                photoURL:
+                  userData?.photoURL || firebaseUser.photoURL || undefined,
+                access: userData.access || "user",
+              };
+              resolve(mappedUser);
+            } else {
+              resolve(null);
+            }
+          } catch (error) {
+            reject(error);
+          }
+        } else {
+          resolve(null);
+        }
+        unsubscribe();
+      },
+      reject,
+    );
+  });
+};
```

Comment:
try to reduce the level of nesting here by separating parts of the logic out. Also, reverse the conditions so you can use guard clauses

example:
```ts
if (condition) {
} else {
  reject(...)
}

// becomes
if (!condition) {
  reject(...)
}
// original code
...
```